pkg/inject/inject.go

```diff
@@ -305,24 +306,32 @@ func (conf *ResourceConfig) GetPodPatch(injectProxy bool) ([]byte, error) {
 	return res, nil
 }
 
-// GetServicePatch returns the JSON patch containing the service opaque ports
-// annotation if the annotation is present on the namespace, but absent on the
-// service.
-func (conf *ResourceConfig) GetServicePatch() ([]byte, error) {
-	_, ok := conf.workload.Meta.Annotations[k8s.ProxyOpaquePortsAnnotation]
-	// There does not need to be a patch if the service already has the
-	// annotation.
+// GetOpaquePorts returns two values. The first value is the the opaque ports
+// value if it is present on the workload or the namespace. The second value
+// is true if it was on the workload and false otherwise.
+func (conf *ResourceConfig) GetOpaquePorts() (string, bool) {
```

Comment:
This API is a bit confusing to me because this method can return the opaque ports of the workload, but we never use that value.  i.e. we only use the first return value when the second return false is false.  I think an API which is easier to understand would be something like:

```
func (conf *ResourceConfig) HasOpaquePorts() bool
func (conf *ResourceConfig) NsOpaquePorts() (string, bool)

if !resourceConfig.HasOpaquePorts() {
 if opaquePorts, ok := resourceConfig.NsOpaquePorts(); ok {
    ...
  }
}
```

Reply:
I agree naming could at least change for this. Maybe something like:

```
func (conf *ResourceConfig) ShouldAnnotate() (string, bool)

opaquePorts, shouldAnnotate := resourceConfig.ShouldAnnotate() {
  if opaquePorts != "" && shouldAnotate {
    resourceConfig.AppendPodAnnotation(pkgK8s.ProxyOpaquePortsAnnotation, annotation)
  }
}
```

The reason I think it should still keep this structure is we need to be able to reference `opaquePorts` and `shouldAnnotate` later in the function if things are reordered like you suggest below.

This is because here, we need to append the pod annotation so that when we check if it is injectable, the annotation has been added. If the the resource is not injectable, then we'd need to check `if opaquePorts != "" && shouldAnotate` again in order to create the annotation patch.

Reply:
Ahhh, I understand now: the `AppendPodAnnotation` is needed for the pod injection case as well.  That makes sense.

I notice that we check both `opaquePorts != ""` AND `shouldAnotate`.  Does this mean that if a workload does not have an opaque ports annotation and the namespace has an opaque ports annotation of "", then the empty annotation won't get copied to the pod?  Is this a supported use-case i.e. wanting to turn off all opaque ports at the namespace level?

Another slightly different API that might be clearer here is:

```
func (conf *ResourceConfig) OpaquePortAnnotationsToCopy() map[string]string


for k, v := range resourceConfig.OpaquePortAnnotationsToCopy() {
  resourceConfig.AppendPodAnnotation(k, v)
}
```

This lets the ResourceConfig decide what to return: it can return an empty map when the workload already has the annotation or if the namespace does not have the annotation.

Reply:
I guess this is a little awkward because `CreateAnnotationPatch` just takes a single value.  Honestly, we're really working around Go's lack of `Option` type here.  I think what you originally suggest is probably the best and most idiomatic Go, but with 1 of these changes:

* If we indeed want to skip copying the annotation if it's empty, then we can just use the empty string as a marker to indicate that we don't want to copy.  In other words: `GetOpaquePorts` should only return a string, but in the case where we don't want to copy (i.e. when the workload has the annotation) then return the empty string.
* On the other hand, if we want to support copying a potentially empty string valued annotation, we can keep the current signature but remove the empty string check.  That makes things look a bit more go-like:

```
opaquePorts, ok := resourceConfig.GetOpaquePorts()
```

Reply:
Yep we should support the `""` annotation so I'll fix that. The second suggestion with just checking `ok`—which essentially means "should we annotate"—seems good to me.

Reply:
:+1: with these semantics I think we should also update the implementation of `GetOpaquePorts` so that it doesn't return a real value and false at the same time.  In other words, if the annotation is already on the workload and we don't need to annotate, we should return `"", false` instead of `annotation, false`.

Reply:
Yep makes sense!